### PR TITLE
Fix final and abstract access modifiers

### LINKED ISSUE #32: Removes FB name from declaration when using FINAL
Removed the FB name. My initial guess is due to the FINAL descriptor.
image

With a FINAL method also weird behavior with an extra colon added.
image

If you point me in the right direction I can take a look at it.

### PR DESCRIPTION
**Pull request by kdorsel**
Fix for abstract and final access modifiers for FB and methods
Fixes #32 #36

### CODE REVIEW

**Comment by Roald87** [comment]
Hey kdorsel,

Thanks you very much for the PR. I'll have look at it tomorrow.

Roald

File: src/TcBlack/ObjectDefinition.cs [hunk 1]
```diff
@@ -138,7 +153,7 @@ private TcObject TokenizeFunctionBlock()
         private TcObject TokenizeMethodOrProperty()
         {
             string entityType = @"\s*(FUNCTION|METHOD|PROPERTY)\s*";
-            string accessModifier = @"(PRIVATE|PUBLIC|PROTECTED|INTERNAL)?\s*";
+            string accessModifier = @"(PRIVATE|PUBLIC|PROTECTED|INTERNAL|FINAL|ABSTRACT)?\s*";
```
Comment (Roald87) [inline src/TcBlack/ObjectDefinition.cs:156]:
```suggestion
            string accessModifier = 
                @"(PRIVATE|PUBLIC|PROTECTED|INTERNAL|FINAL|ABSTRACT)?\s*";
```

I'm trying to also maintain the 88 line length limit in the code itself. See for more info https://github.com/Roald87/TcBlack/blob/master/CONTRIBUTING.md#use-a-consistent-coding-style

Comment (kdorsel) [reply src/TcBlack/ObjectDefinition.cs]: Sure, is there a C# Black that can be added to the project?

**Comment by kdorsel** [comment]
> Can you also add some tests to the TcBlackTests project in `ObjectDefinitionTests.cs`? Perhaps best to collect the tests in a new method, something like `AbstractAndFinalModifiersForMethodsAndFunctionBlocks`

I'll create a test suite for these.

**Comment by kdorsel** [comment]
I'm having troubles with loading the test project with this error:

`C:\Users\kdorsel\Documents\github\TcBlack\src\TcBlackTests\TcBlackTests.csproj : error  : The default XML namespace of the project must be the MSBuild XML namespace. If the project is authored in the MSBuild 2003 format, please add xmlns="http://schemas.microsoft.com/developer/msbuild/2003" to the <Project> element. If the project has been authored in the old 1.0 or 1.2 format, please convert it to MSBuild 2003 format.  C:\Users\kdorsel\Documents\github\TcBlack\src\TcBlackTests\TcBlackTests.csproj`


Comparing the TcBlack and TcBlackTests csproj they are completely different. I'm using VS2015.

File: src/TcBlack/ObjectDefinition.cs (hunk 1 above)
Comment (Roald87) [reply src/TcBlack/ObjectDefinition.cs]: I haven't found a free one unfortunately. There is https://www.jetbrains.com/resharper/, but you have to pay for that one.

**Comment by Roald87** [comment]
I haven't tried to run the project with VS2015. In the contributing https://github.com/Roald87/TcBlack/blob/master/CONTRIBUTING.md#use-a-consistent-coding-style document I recommend to use VS2017. Is that an issue for you?

The fact that they are so different is an artifact from the past. I initially stared the project using .NET Core. Later I converted to the .NET Framework. Apparently I didn't convert the unit test project.

I'll have a look if I can convert the unit test project as well.

**Comment by kdorsel** [comment]
I need VS2015 since TC3 Database Server doesn't work on VS2017 yet. I can get VS2017 also, but beside that file I haven't had any issues.

**Comment by Roald87** [comment]
Ok let me try to change it. As long as you don't upload the `.csproj` file maybe it'll be fine.

**Comment by Roald87** [comment]
Does the test in this branch run?

https://github.com/Roald87/TcBlack/tree/xunit-in-net-framework

**Comment by kdorsel** [comment]
The UnitTestDotNetFramework project loads and I can access it. But when I Run All Tests I get no results and it looks like no tests are run.

I'll look more into it. I'll also get VS2017 and try and report back.

https://xunit.net/docs/getting-started/netfx/visual-studio
https://stackoverflow.com/questions/44573009/does-xunit-runner-visualstudio-not-work-for-visual-studio

**Comment by Roald87** [comment]
Is was some option under Tests > General I believe which you have to disable. It was something with optimization blabla. You should turn that off, then it worked for me.

**Comment by kdorsel** [comment]
In VS2017, everything works fine from the main branch (TcBlackTests). I'll work on adding the tests now for this.

Sorry for adding the extra work! The xunit-in-net-framework branch is no longer needed.

**Comment by Roald87** [comment]
No problem. I didn't want to scare away the first contributor 😛 .

**Comment by kdorsel** [comment]
Updated with tests and also closes #36 now.

File: src/TcBlack/ObjectDefinition.cs [hunk 2]
```diff
@@ -124,25 +124,19 @@ private TcObject TokenizeFunctionBlock()
                     extends = true;
                     implements = false;
                 }
-                else if (part.ToLower() == "abstract" || part.ToLower() == "final")
+                else if (part.ToLower() == "abstract"
+                    || part.ToLower() == "final"
+                    || part.ToLower() == "internal"
+                    || part.ToLower() == "public")
```
Comment (Roald87) [inline src/TcBlack/ObjectDefinition.cs:130]:
```suggestion
                else if (
                    part.ToLower() == "abstract"
                    || part.ToLower() == "final"
                    || part.ToLower() == "internal"
                    || part.ToLower() == "public"
                )
```

File: src/TcBlack/ObjectDefinition.cs [hunk 3]
```diff
@@ -97,7 +97,7 @@ private TcObject TokenizeFunctionBlock()
 
             List<string> interfaces = new List<string>();
             List<string> parents = new List<string>();
-            string accessModifier = "";
+            List<string> accessModifier = new List<string>();
```
Comment (Roald87) [inline src/TcBlack/ObjectDefinition.cs:100]:
```suggestion
            List<string> accessModifiers = new List<string>();
```

File: src/TcBlack/ObjectDefinition.cs (hunk 2 above)
Comment (Roald87) [reply src/TcBlack/ObjectDefinition.cs:130]: Sorry for the small changes, but if your building a code formatter the code of the formatter itself should also be consistent :p.

**Comment by Roald87** [comment]
Thank you very much for your contribution!

**Pull request merged**